In [1]:
import pandas as pd
import json
import numpy as np

In [2]:
json_file = "RegulonDB/RegulonDB_v12/regulondbdatamarts.operonDatamart.json"

with open(json_file, 'r') as f:
    datos = json.load(f)



In [134]:
## no usar

df = pd.DataFrame({"Operon ID":[] , 
                   "Operon name":[], 
                   "Operon strand": [], 
                   "TU ID": [], 
                   "TU name":[],
                   "TU first gene":[],                    
                   "Promoter ID":[], 
                   "Promoter name":[],                   
                   "Sigma factor":[], 
                   "Promoter sequence":[],
                   "Promoter -35 box start": [],
                   "Promoter -35 box end": [],
                   "Promoter -35 box sequence": [],
                   "Promoter -10 box start": [],
                   "Promoter -10 box end": [],
                   "Promoter -10 box sequence": [],                                 
                   "Promoter confidence": []}, dtype='object')

row = 0
for i in range(len(datos)):
    
    for elem2 in datos[i]['transcriptionUnits']:
        df.loc[row, "Operon ID"] = datos[i]['_id']
        df.loc[row, "Operon name"] = datos[i]['operon']['name']
        df.loc[row, "Operon strand"] = datos[i]['operon']['strand']
        df.loc[row, "TU ID"] = elem2['_id']
        try:
            df.loc[row, "TU name"] = elem2['name']
            df.loc[row, "TU first gene"] = elem2['firstGene']['name']            
            df.loc[row, "Promoter ID"] = elem2['promoter']['_id']
            df.loc[row, "Promoter name"] = elem2['promoter']['name']            
            df.loc[row, "Sigma factor"] = elem2['promoter']['bindsSigmaFactor']['abbreviatedName']
            df.loc[row, "Promoter sequence"] = elem2['promoter']['sequence']            
            df.loc[row, "Promoter -35 box start"] = elem2['promoter']['boxes'][1]["leftEndPosition"]
            df.loc[row, "Promoter -35 box end"] = elem2['promoter']['boxes'][1]["rightEndPosition"]
            df.loc[row, "Promoter -35 box sequence"] = elem2['promoter']['boxes'][1]["sequence"]
            df.loc[row, "Promoter -10 box start"] = elem2['promoter']['boxes'][0]["leftEndPosition"]
            df.loc[row, "Promoter -10 box end"] = elem2['promoter']['boxes'][0]["rightEndPosition"]
            df.loc[row, "Promoter -10 box sequence"] = elem2['promoter']['boxes'][0]["sequence"]
            df.loc[row, "Promoter confidence"] = elem2['promoter']['confidenceLevel']
    
            row = row + 1
        except:
            row = row + 1
            continue
        


In [3]:
import pandas as pd

# Crear el DataFrame con las columnas deseadas
df = pd.DataFrame({"Operon ID":[], 
                   "Operon name":[], 
                   "Operon strand": [], 
                   "TU ID": [], 
                   "TU name":[],
                   "TU first gene":[],                    
                   "Promoter ID":[], 
                   "Promoter name":[],                   
                   "Sigma factor":[], 
                   "Promoter sequence":[],
                   "Promoter -35 box start": [],
                   "Promoter -35 box end": [],
                   "Promoter -35 box sequence": [],
                   "Promoter -10 box start": [],
                   "Promoter -10 box end": [],
                   "Promoter -10 box sequence": [],                                 
                   "Promoter confidence": []}, dtype='object')

# Inicializar la fila
row = 0

# Iterar sobre los datos
for i in range(len(datos)):
    for elem2 in datos[i]['transcriptionUnits']:
        df.loc[row, "Operon ID"] = datos[i]['_id']
        df.loc[row, "Operon name"] = datos[i]['operon']['name']
        df.loc[row, "Operon strand"] = datos[i]['operon']['strand']
        df.loc[row, "TU ID"] = elem2['_id']
        
        try:
            try:
                df.loc[row, "TU name"] = elem2['name']
            except KeyError:
                pass

            try:
                df.loc[row, "TU first gene"] = elem2['firstGene']['name']
            except KeyError:
                pass

            try:
                df.loc[row, "Promoter ID"] = elem2['promoter']['_id']
            except KeyError:
                pass

            try:
                df.loc[row, "Promoter name"] = elem2['promoter']['name']
            except KeyError:
                pass

            try:
                df.loc[row, "Sigma factor"] = elem2['promoter']['bindsSigmaFactor']['abbreviatedName']
            except KeyError:
                pass

            try:
                df.loc[row, "Promoter sequence"] = elem2['promoter']['sequence']
            except KeyError:
                pass

            try:
                df.loc[row, "Promoter -35 box start"] = elem2['promoter']['boxes'][1]["leftEndPosition"]
            except (KeyError, IndexError):
                pass

            try:
                df.loc[row, "Promoter -35 box end"] = elem2['promoter']['boxes'][1]["rightEndPosition"]
            except (KeyError, IndexError):
                pass

            try:
                df.loc[row, "Promoter -35 box sequence"] = elem2['promoter']['boxes'][1]["sequence"]
            except (KeyError, IndexError):
                pass

            try:
                df.loc[row, "Promoter -10 box start"] = elem2['promoter']['boxes'][0]["leftEndPosition"]
            except (KeyError, IndexError):
                pass

            try:
                df.loc[row, "Promoter -10 box end"] = elem2['promoter']['boxes'][0]["rightEndPosition"]
            except (KeyError, IndexError):
                pass

            try:
                df.loc[row, "Promoter -10 box sequence"] = elem2['promoter']['boxes'][0]["sequence"]
            except (KeyError, IndexError):
                pass

            try:
                df.loc[row, "Promoter confidence"] = elem2['promoter']['confidenceLevel']
            except KeyError:
                pass

        except Exception as e:
            print(f"Error procesando la fila {row}: {e}")

        row += 1


In [4]:
df       

,Operon ID,Operon name,Operon strand,TU ID,TU name,TU first gene,Promoter ID,Promoter name,Sigma factor,Promoter sequence,Promoter -35 box start,Promoter -35 box end,Promoter -35 box sequence,Promoter -10 box start,Promoter -10 box end,Promoter -10 box sequence,Promoter confidence
0,RDBECOLIOPC00001,ispDF,reverse,RDBECOLITUC01454,ispDF,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,RDBECOLIOPC00002,serX,reverse,RDBECOLITUC03528,serX,serX,RDBECOLIPMC03869,serXp,sigma70,gcgattttttcaaaaagtttgttgacctcaggtcatgatttcccta...,1097710,1097715,TTGACC,1097687,1097692,TAAATT,W
2,RDBECOLIOPC00003,ybjS,reverse,RDBECOLITUC00341,ybjS,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,RDBECOLIOPC00004,insJ,forward,RDBECOLITUC01510,insJ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,RDBECOLIOPC00005,murJ,forward,RDBECOLITUC00397,murJ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3737,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02233,pstSCA,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600,3911605,CATATA,3911576,3911586,TGTCCTATTTT,S
3738,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02234,pstS,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600,3911605,CATATA,3911576,3911586,TGTCCTATTTT,S
3739,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02235,pstSCAB-phoU,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600,3911605,CATATA,3911576,3911586,TGTCCTATTTT,S
3740,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02258,pstB-phoU,pstB,RDBECOLIPMC02689,pstBp2,sigma38,ttaccctgtgcgtactgctgctgaacattctggcgcgcgttgtttt...,3908582,3908587,CTGGCG,3908561,3908571,TTGCGAAGAAT,S


In [61]:
df.to_csv("dataframes/2_regulonDB_parsed.csv", index=False)

In [181]:
df = pd.read_csv("dataframes/2_regulonDB_parsed.csv")

In [182]:
df

,Operon ID,Operon name,Operon strand,TU ID,TU name,TU first gene,Promoter ID,Promoter name,Sigma factor,Promoter sequence,Promoter -35 box start,Promoter -35 box end,Promoter -35 box sequence,Promoter -10 box start,Promoter -10 box end,Promoter -10 box sequence,Promoter confidence
0,RDBECOLIOPC00001,ispDF,reverse,RDBECOLITUC01454,ispDF,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,RDBECOLIOPC00002,serX,reverse,RDBECOLITUC03528,serX,serX,RDBECOLIPMC03869,serXp,sigma70,gcgattttttcaaaaagtttgttgacctcaggtcatgatttcccta...,1097710.0,1097715.0,TTGACC,1097687.0,1097692.0,TAAATT,W
2,RDBECOLIOPC00003,ybjS,reverse,RDBECOLITUC00341,ybjS,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,RDBECOLIOPC00004,insJ,forward,RDBECOLITUC01510,insJ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,RDBECOLIOPC00005,murJ,forward,RDBECOLITUC00397,murJ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3737,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02233,pstSCA,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S
3738,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02234,pstS,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S
3739,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02235,pstSCAB-phoU,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S
3740,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02258,pstB-phoU,pstB,RDBECOLIPMC02689,pstBp2,sigma38,ttaccctgtgcgtactgctgctgaacattctggcgcgcgttgtttt...,3908582.0,3908587.0,CTGGCG,3908561.0,3908571.0,TTGCGAAGAAT,S


# Check that all the promoter_ids are present

In [183]:
df_promoters = pd.read_csv("RegulonDB/RegulonDB_v12/promoters_id.csv")

In [184]:
df_promoters

,_id,transcriptionUnits[0].promoter._id,transcriptionUnits[1].promoter._id,transcriptionUnits[2].promoter._id,transcriptionUnits[3].promoter._id,transcriptionUnits[4].promoter._id,transcriptionUnits[5].promoter._id,transcriptionUnits[6].promoter._id,transcriptionUnits[7].promoter._id,transcriptionUnits[8].promoter._id,...,transcriptionUnits[9].promoter.boxes[0].sequence,transcriptionUnits[9].promoter.boxes[1].sequence,transcriptionUnits[10].promoter.boxes[0].sequence,transcriptionUnits[10].promoter.boxes[1].sequence,transcriptionUnits[12].promoter.boxes[0].sequence,transcriptionUnits[12].promoter.boxes[1].sequence,transcriptionUnits[13].promoter.boxes[0].sequence,transcriptionUnits[13].promoter.boxes[1].sequence,transcriptionUnits[11].promoter.boxes[0].sequence,transcriptionUnits[11].promoter.boxes[1].sequence
0,RDBECOLIOPC00001,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,RDBECOLIOPC00002,RDBECOLIPMC03869,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,RDBECOLIOPC00003,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,RDBECOLIOPC00004,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,RDBECOLIOPC00005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2608,RDBECOLIOPC03555,RDBECOLIPMC02599,RDBECOLIPMC02600,RDBECOLIPMC03653,RDBECOLIPMC03651,RDBECOLIPMC03652,RDBECOLIPMC03652,RDBECOLIPMC03876,RDBECOLIPMC02961,RDBECOLIPMC03653,...,TATAAT,TTGACA,TAGAAT,TTGCAT,NaN,NaN,NaN,NaN,NaN,NaN
2609,RDBECOLIOPC03556,RDBECOLIPMC03147,NaN,RDBECOLIPMC02678,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2610,RDBECOLIOPC03557,RDBECOLIPMC03674,RDBECOLIPMC03673,RDBECOLIPMC03672,RDBECOLIPMC03671,RDBECOLIPMC03669,RDBECOLIPMC03670,RDBECOLIPMC02637,RDBECOLIPMC02648,RDBECOLIPMC02722,...,ACTTAG,NaN,TATGAT,AGATCT,NaN,NaN,NaN,NaN,NaN,NaN
2611,RDBECOLIOPC03558,NaN,RDBECOLIPMC00251,RDBECOLIPMC00252,RDBECOLIPMC02663,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [185]:
L = []
for i in range(1,15):
    for elem in df_promoters.iloc[:,i]:
        if pd.isnull(elem):
            continue
        else:
            L.append(elem)

In [186]:
len(set(L))

2281

In [187]:
promoters_df_L = [x for x in df["Promoter ID"] if not pd.isnull(x)]

In [188]:
len(set(promoters_df_L))

2281

In [189]:
missing_prom = []
for elem in set(L):
    if elem not in set(promoters_df_L):
        missing_prom.append(elem)

In [190]:
missing_prom

[]

# Find the spacer region of promoters

In [191]:
from Bio import SeqIO
from Bio.Seq import Seq

# Ruta al archivo FASTA
fasta_file = "K12\\NC_000913.fasta"

# Leer el archivo FASTA que contiene una única secuencia
for seq_record in SeqIO.parse(fasta_file, "fasta"):
    K12_genome = str(seq_record.seq)

In [192]:
def extract_spacer(start35, end35, start10, end10, strand, genome):

    try:
        if strand == "forward":
            spacer = genome[int(end35):int(start10)-1]

        if strand == "reverse":
            spacer = genome[int(end10):int(start35)-1]
            seq = Seq(spacer)
            spacer = str(seq.reverse_complement())

        return spacer
    except:
        return ""

In [193]:
df.head(1)

,Operon ID,Operon name,Operon strand,TU ID,TU name,TU first gene,Promoter ID,Promoter name,Sigma factor,Promoter sequence,Promoter -35 box start,Promoter -35 box end,Promoter -35 box sequence,Promoter -10 box start,Promoter -10 box end,Promoter -10 box sequence,Promoter confidence
0,RDBECOLIOPC00001,ispDF,reverse,RDBECOLITUC01454,ispDF,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [194]:
df["spacer"] = df.apply(lambda row: extract_spacer(row["Promoter -35 box start"], row["Promoter -35 box end"], row["Promoter -10 box start"], row["Promoter -10 box end"], row["Operon strand"], K12_genome), axis=1)

In [195]:
df

,Operon ID,Operon name,Operon strand,TU ID,TU name,TU first gene,Promoter ID,Promoter name,Sigma factor,Promoter sequence,Promoter -35 box start,Promoter -35 box end,Promoter -35 box sequence,Promoter -10 box start,Promoter -10 box end,Promoter -10 box sequence,Promoter confidence,spacer
0,RDBECOLIOPC00001,ispDF,reverse,RDBECOLITUC01454,ispDF,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,
1,RDBECOLIOPC00002,serX,reverse,RDBECOLITUC03528,serX,serX,RDBECOLIPMC03869,serXp,sigma70,gcgattttttcaaaaagtttgttgacctcaggtcatgatttcccta...,1097710.0,1097715.0,TTGACC,1097687.0,1097692.0,TAAATT,W,TCAGGTCATGATTTCCC
2,RDBECOLIOPC00003,ybjS,reverse,RDBECOLITUC00341,ybjS,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,
3,RDBECOLIOPC00004,insJ,forward,RDBECOLITUC01510,insJ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,
4,RDBECOLIOPC00005,murJ,forward,RDBECOLITUC00397,murJ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3737,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02233,pstSCA,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT
3738,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02234,pstS,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT
3739,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02235,pstSCAB-phoU,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT
3740,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02258,pstB-phoU,pstB,RDBECOLIPMC02689,pstBp2,sigma38,ttaccctgtgcgtactgctgctgaacattctggcgcgcgttgtttt...,3908582.0,3908587.0,CTGGCG,3908561.0,3908571.0,TTGCGAAGAAT,S,CGCGTTGTTT


## Check promoters sequences are OK

In [196]:
df["promoter sequence reduced"] = df.apply(lambda row: str(row["Promoter -35 box sequence"]) + str(row["spacer"]) + str(row["Promoter -10 box sequence"]), axis=1)

In [197]:
df

,Operon ID,Operon name,Operon strand,TU ID,TU name,TU first gene,Promoter ID,Promoter name,Sigma factor,Promoter sequence,Promoter -35 box start,Promoter -35 box end,Promoter -35 box sequence,Promoter -10 box start,Promoter -10 box end,Promoter -10 box sequence,Promoter confidence,spacer,promoter sequence reduced
0,RDBECOLIOPC00001,ispDF,reverse,RDBECOLITUC01454,ispDF,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,,nannan
1,RDBECOLIOPC00002,serX,reverse,RDBECOLITUC03528,serX,serX,RDBECOLIPMC03869,serXp,sigma70,gcgattttttcaaaaagtttgttgacctcaggtcatgatttcccta...,1097710.0,1097715.0,TTGACC,1097687.0,1097692.0,TAAATT,W,TCAGGTCATGATTTCCC,TTGACCTCAGGTCATGATTTCCCTAAATT
2,RDBECOLIOPC00003,ybjS,reverse,RDBECOLITUC00341,ybjS,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,,nannan
3,RDBECOLIOPC00004,insJ,forward,RDBECOLITUC01510,insJ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,,nannan
4,RDBECOLIOPC00005,murJ,forward,RDBECOLITUC00397,murJ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,,nannan
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3737,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02233,pstSCA,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT,CATATAACTGTCACCTGTTTGTCCTATTTT
3738,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02234,pstS,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT,CATATAACTGTCACCTGTTTGTCCTATTTT
3739,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02235,pstSCAB-phoU,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT,CATATAACTGTCACCTGTTTGTCCTATTTT
3740,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02258,pstB-phoU,pstB,RDBECOLIPMC02689,pstBp2,sigma38,ttaccctgtgcgtactgctgctgaacattctggcgcgcgttgtttt...,3908582.0,3908587.0,CTGGCG,3908561.0,3908571.0,TTGCGAAGAAT,S,CGCGTTGTTT,CTGGCGCGCGTTGTTTTTGCGAAGAAT


In [198]:
df["promoter sequence reduced"] = df["promoter sequence reduced"].replace("nannan", "")

In [199]:
df["promoter sequence reduced"] = df["promoter sequence reduced"].apply(lambda x: "" if x.startswith("nan") else x)

In [200]:
df

,Operon ID,Operon name,Operon strand,TU ID,TU name,TU first gene,Promoter ID,Promoter name,Sigma factor,Promoter sequence,Promoter -35 box start,Promoter -35 box end,Promoter -35 box sequence,Promoter -10 box start,Promoter -10 box end,Promoter -10 box sequence,Promoter confidence,spacer,promoter sequence reduced
0,RDBECOLIOPC00001,ispDF,reverse,RDBECOLITUC01454,ispDF,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,,
1,RDBECOLIOPC00002,serX,reverse,RDBECOLITUC03528,serX,serX,RDBECOLIPMC03869,serXp,sigma70,gcgattttttcaaaaagtttgttgacctcaggtcatgatttcccta...,1097710.0,1097715.0,TTGACC,1097687.0,1097692.0,TAAATT,W,TCAGGTCATGATTTCCC,TTGACCTCAGGTCATGATTTCCCTAAATT
2,RDBECOLIOPC00003,ybjS,reverse,RDBECOLITUC00341,ybjS,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,,
3,RDBECOLIOPC00004,insJ,forward,RDBECOLITUC01510,insJ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,,
4,RDBECOLIOPC00005,murJ,forward,RDBECOLITUC00397,murJ,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,,
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3737,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02233,pstSCA,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT,CATATAACTGTCACCTGTTTGTCCTATTTT
3738,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02234,pstS,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT,CATATAACTGTCACCTGTTTGTCCTATTTT
3739,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02235,pstSCAB-phoU,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,3911600.0,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT,CATATAACTGTCACCTGTTTGTCCTATTTT
3740,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02258,pstB-phoU,pstB,RDBECOLIPMC02689,pstBp2,sigma38,ttaccctgtgcgtactgctgctgaacattctggcgcgcgttgtttt...,3908582.0,3908587.0,CTGGCG,3908561.0,3908571.0,TTGCGAAGAAT,S,CGCGTTGTTT,CTGGCGCGCGTTGTTTTTGCGAAGAAT


### Check if promote sequence reduced are in promoter sequences

In [201]:
for i in range(len(df.index)):
    if str(df["promoter sequence reduced"][i]).lower() not in str(df["Promoter sequence"][i]):
        print(i, df["promoter sequence reduced"][i])

912 TCGAAAAATAGCCCGCTGATATCAT
1764 TCTCAAATCATA
2377 TTGAAGACGTTTGGGTGTTCACCAACGT
3342 TTAAATCTGTTA


912 TCGAAAAATAGCCCGCTGATATCAT does not have the Promoter sequence, confidenceLevel : W


1764 TCTCAAATCATA, is a weird promoter, it is in a invertible region of the genome. I modify mannually

In [202]:
df.loc[1764,"Promoter -35 box sequence"] = df.loc[1764,"Promoter -35 box sequence"][::-1]

In [203]:
df.loc[1764,"Promoter -10 box sequence"] = df.loc[1764,"Promoter -10 box sequence"][::-1]

In [204]:
df.loc[1764,"spacer"] = "aataagttacgtattttt"[::-1]

In [205]:
df.loc[1764,"Promoter sequence"] = df.loc[1764,"Promoter sequence"][::-1]

In [206]:
df.loc[1764,"promoter sequence reduced"] = "atcataaataagttacgtattttttctcaa"[::-1]

In [209]:
for i in range(len(df.index)):
    if str(df["promoter sequence reduced"][i]).lower() not in str(df["Promoter sequence"][i]):
        print(i, df["promoter sequence reduced"][i])

912 TCGAAAAATAGCCCGCTGATATCAT
2377 TTGAAGACGTTTGGGTGTTCACCAACGT
3342 TTAAATCTGTTA


    2377 TTGAAGACGTTTGGGTGTTCACCAACGT, the coordinates of the boxes are outside the promoter sequence. It does not have sigma factor. Confidence Level W
    3342 TTAAATCTGTTA, is an outlier promoter. The -10 and -35 boxes are overlapped. Confidence Level S

In [210]:
df.to_csv("dataframes/3_regulonDB_parsed_and_checked.csv", index=False)

# Adding locus_tag to the dataframe

In [2]:
df = pd.read_csv("dataframes/3_regulonDB_parsed_and_checked.csv")

In [11]:
genes_df_regulonDB = pd.read_csv("K12/GeneProductSet.txt", sep="\t", skiprows=41, header=None)

In [20]:
df = pd.merge(df, genes_df_regulonDB[[1,2]], left_on="TU first gene", right_on=1)

In [21]:
df

,Operon ID,Operon name,Operon strand,TU ID,TU name,TU first gene,Promoter ID,Promoter name,Sigma factor,Promoter sequence,...,Promoter -35 box end,Promoter -35 box sequence,Promoter -10 box start,Promoter -10 box end,Promoter -10 box sequence,Promoter confidence,spacer,promoter sequence reduced,1,2
0,RDBECOLIOPC00001,ispDF,reverse,RDBECOLITUC01454,ispDF,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,RDBECOLIOPC00002,serX,reverse,RDBECOLITUC03528,serX,serX,RDBECOLIPMC03869,serXp,sigma70,gcgattttttcaaaaagtttgttgacctcaggtcatgatttcccta...,...,1097715.0,TTGACC,1097687.0,1097692.0,TAAATT,W,TCAGGTCATGATTTCCC,TTGACCTCAGGTCATGATTTCCCTAAATT,serX,b1032
2,RDBECOLIOPC00003,ybjS,reverse,RDBECOLITUC00341,ybjS,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,RDBECOLIOPC00004,insJ,forward,RDBECOLITUC01510,insJ,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,RDBECOLIOPC00005,murJ,forward,RDBECOLITUC00397,murJ,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3710,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02233,pstSCA,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,...,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT,CATATAACTGTCACCTGTTTGTCCTATTTT,pstS,b3728
3711,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02234,pstS,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,...,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT,CATATAACTGTCACCTGTTTGTCCTATTTT,pstS,b3728
3712,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02235,pstSCAB-phoU,pstS,RDBECOLIPMC02666,pstSp2,sigma38,tcataaaactgtcatattccttacatataactgtcacctgtttgtc...,...,3911605.0,CATATA,3911576.0,3911586.0,TGTCCTATTTT,S,ACTGTCACCTGTT,CATATAACTGTCACCTGTTTGTCCTATTTT,pstS,b3728
3713,RDBECOLIOPC03559,pstSCAB-phoU,reverse,RDBECOLITUC02258,pstB-phoU,pstB,RDBECOLIPMC02689,pstBp2,sigma38,ttaccctgtgcgtactgctgctgaacattctggcgcgcgttgtttt...,...,3908587.0,CTGGCG,3908561.0,3908571.0,TTGCGAAGAAT,S,CGCGTTGTTT,CTGGCGCGCGTTGTTTTTGCGAAGAAT,pstB,b3725


In [24]:
df.drop(columns=1, inplace=True)

In [27]:
df.rename(columns={2:"locus tag"}, inplace=True)

In [28]:
df.to_csv("dataframes/8_regulonDB_parsed_and_checked_locusTag.csv", index=False)